### PR TITLE
[mac build] Build early swift-driver

### PR DESCRIPTION
**Pull request by Steelskin**
This is a required preliminary step to build the compiler on macOS.

* Separate Swift flags between Windows and Darwin.
* Move arch-specific and OS-specific flags to matrix.extra_flags.

### CODE REVIEW

File: .github/workflows/swift-toolchain.yml [hunk 1]
```diff
@@ -700,6 +698,320 @@ jobs:
...
+      - uses: compnerd/gha-setup-vsdevenv@main
+        with:
+          host_arch: ${{ inputs.build_arch }}
+          components: 'Microsoft.VisualStudio.Component.VC.Tools.x86.x64;Microsoft.VisualStudio.Component.VC.Tools.ARM64'
+          arch: ${{ matrix.arch }}
+
+      - uses: seanmiddleditch/gha-setup-ninja@master
+        if: inputs.build_os == 'Darwin'
+
+      # TODO: Use the TBC builds once we have a Mac toolchain release.
```
[the remark this answers is not in the document]
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml]: Dropped the TODO

File: .github/workflows/swift-toolchain.yml [hunk 2]
```diff
@@ -700,6 +698,320 @@ jobs:
...
+          $InstallFolderName = "earlyswiftdriver-${PlatformName}-${{ matrix.arch }}"
+
+          echo "SDKROOT=$SDKROOT" >> $env:GITHUB_OUTPUT
+          echo "SQLite3_lib=$SQLite3_lib" >> $env:GITHUB_OUTPUT
+          echo "InstallFolderName=$InstallFolderName" >> $env:GITHUB_OUTPUT
+
+      - name: Configure swift-llbuild
+        run: |
+          cmake -B ${{ github.workspace }}/BinaryCache/swift-llbuild `
+                -D BUILD_SHARED_LIBS=YES `
```
[the remark this answers is not in the document]
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml:790]: We'll have to deal with the rpath either way. Right now, most of these do not build properly as static libraries. I am open to doing the changes needed upstream to make it happen, though. At the same time, I could fix the installation paths so we don't have to do this dance for installation.
How would you like to proceed?

File: .github/workflows/swift-toolchain.yml [hunk 3]
```diff
@@ -700,6 +698,320 @@ jobs:
...
+            $SDKROOT = "${env:SDKROOT}"
+            $SQLite3_lib = "${{ github.workspace }}/BuildRoot/Library/sqlite-${{ inputs.swift_toolchain_sqlite_version }}/usr/lib/SQLite3.lib"
+            $PlatformName = "windows"
+          } else {
+            $SDKROOT = xcrun --sdk macosx --show-sdk-path
+            $SQLite3_lib = "${{ github.workspace }}/BuildRoot/Library/sqlite-${{ inputs.swift_toolchain_sqlite_version }}/usr/lib/libSQLite3.a"
+            $PlatformName = "macosx"
+          }
+
+          # The Swift compiler expects a specific folder name for the early swift-driver build.
```
[the remark this answers is not in the document]
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml:779]: I can try to get that landed if you'd like. In the meantime, we do need a specific directory. I added a TODO for now.

File: .github/workflows/swift-toolchain.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml]: As discussed offline, I switched to using SPM here.

File: .github/workflows/swift-toolchain.yml [hunk 4]
```diff
@@ -700,6 +698,123 @@ jobs:
...
+        with:
+          host_arch: ${{ inputs.build_arch }}
+          components: 'Microsoft.VisualStudio.Component.VC.Tools.x86.x64;Microsoft.VisualStudio.Component.VC.Tools.ARM64'
+          arch: ${{ matrix.arch }}
+
+      - uses: seanmiddleditch/gha-setup-ninja@master
+        if: inputs.build_os == 'Darwin'
+
+      - name: Install Swift Toolchain (Mac)
+        if: inputs.build_os == 'Darwin'
```
Comment (hjyamauchi) [inline .github/workflows/swift-toolchain.yml:760]: On Windows, where do we install swift?

File: .github/workflows/swift-toolchain.yml [hunk 5]
```diff
@@ -700,6 +698,123 @@ jobs:
...
+          path: ${{ github.workspace }}/SourceCache/Yams
+          show-progress: false
+      - uses: actions/checkout@v4
+        with:
+          repository: swiftlang/swift-driver
+          ref: ${{ inputs.swift_driver_revision }}
+          path: ${{ github.workspace }}/SourceCache/swift-driver
+          show-progress: false
+
+      - uses: compnerd/gha-setup-vsdevenv@main
```
Comment (hjyamauchi) [inline .github/workflows/swift-toolchain.yml:750]: Is it safe to unconditionally have this on Darwin?
IOW, do we want to add `if: inputs.build_os == 'Darwin'` for `gha-setup-vsdevenv`?

File: .github/workflows/swift-toolchain.yml [hunk 6]
```diff
@@ -700,6 +698,123 @@ jobs:
...
+          swift build `
+                --configuration release `
+                --arch ${{ matrix.arch }} `
+                --package-path ${{ github.workspace }}/SourceCache/swift-driver `
+                --build-path ${{ github.workspace }}/BinaryCache/swift-driver `
+
+      - name: Create installable package
+        run: |
+          if ("${{ matrix.os }}" -eq "Darwin") {
+              $PlatformName = "windows"
```
Comment (hjyamauchi) [inline .github/workflows/swift-toolchain.yml:777]: Should this be `$PlatformName = "macosx"`?

File: .github/workflows/swift-toolchain.yml [hunk 7]
```diff
@@ -700,6 +698,123 @@ jobs:
...
+              $ExeSuffix = ""
+              $Cpu = "${{ matrix.arch }}"
+              $SourceBinDir = Join-Path "${{ github.workspace }}" "BinaryCache" "swift-driver" "${Cpu}-apple-macosx" "release"
+          } else {
+              $PlatformName = "macosx"
```
Comment (hjyamauchi) [inline .github/workflows/swift-toolchain.yml:782]: Should this be `$PlatformName = "windows"`?

File: .github/workflows/swift-toolchain.yml (hunk 5 above)
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml:750]: This came up before, @compnerd said that this was a no-op on Mac and we should leave it as-is. See this [discussion](https://github.com/compnerd/swift-build/pull/806#discussion_r1722087115)

File: .github/workflows/swift-toolchain.yml (hunk 4 above)
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml:760]: This job does not have a Windows version yet. @compnerd is working on bringing the early swift-driver build separately.

File: .github/workflows/swift-toolchain.yml (hunk 7 above)
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml]: Yes, I failed to merge the fix on the proper branch.

File: .github/workflows/swift-toolchain.yml (hunk 6 above)
Comment (Steelskin) [reply .github/workflows/swift-toolchain.yml]: Done

**Review by hjyamauchi — APPROVED** [review]

**Pull request merged**